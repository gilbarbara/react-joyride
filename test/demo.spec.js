import React from 'react';
import { mount } from 'enzyme';

import Demo from '../demo/src/App';

const mockConsole = jest.fn();
console.warn = mockConsole; //eslint-disable-line no-console

const mockCallback = jest.fn();
const props = {
  callback: mockCallback,
  run: false
};

function setup(ownProps = props) {
  return mount(
    <Demo joyride={ownProps} />,
    { attachTo: document.getElementById('react') }
  );
}

describe('Joyride', () => {
  let wrapper;

  beforeAll(() => {
    Element.prototype.clientHeight = 36;
    Element.prototype.clientWidth = 36;

    Element.prototype.getBoundingClientRect = jest.fn(() =>
      ({
        width: 1024,
        height: 600,
        top: 450,
        left: 15,
        bottom: 200,
        right: 100,
      })
    );

    const style = document.createElement('style');
    style.type = 'text/css';
    style.innerHTML = '.demo__footer a { width: 100px; height: 60px; }';
    document.body.appendChild(style);
  });

  describe('standalone tooltips', () => {
    wrapper = setup();

    afterAll(() => {
      wrapper.detach();
    });

    it('should be able to trigger the 1st tooltip', () => {
      wrapper.find('.hero__tooltip').get(0).dispatchEvent(new Event('click', { bubbles: true }));

      expect(wrapper.find('.joyride-tooltip--standalone').html()).toMatchSnapshot();
    });

    it('should be able to hide the 1st tooltip', () => {
      wrapper.find('.hero__tooltip').get(0).dispatchEvent(new Event('click', { bubbles: true }));

      expect(wrapper.find('.joyride-tooltip--standalone').length).toBe(0);
    });

    it('should be able to trigger the 2nd tooltip', () => {
      wrapper.find('.demo__footer img').get(0).dispatchEvent(new Event('click', { bubbles: true }));

      expect(wrapper.find('.joyride-tooltip--standalone').html()).toMatchSnapshot();
    });

    it('should be able to hide the 2nd tooltip', () => {
      wrapper.find('.demo__footer img').get(0).dispatchEvent(new Event('mouseleave', { bubbles: true }));

      expect(wrapper.find('.joyride-tooltip--standalone').length).toBe(0);
    });
  });

  describe('tour with `run` set to false', () => {
    beforeAll(() => {
      wrapper = setup();
    });

    afterAll(() => {
      wrapper.detach();
    });

    it('should render properly', () => {
      window.dispatchEvent(new Event('touchstart'));

      expect(wrapper.find('.joyride').length).toEqual(1);
      expect(wrapper.find('.demo__footer').length).toEqual(1);
      expect(wrapper.find('.hero').length).toEqual(1);
      expect(wrapper.find('.projects').length).toEqual(1);
      expect(wrapper.find('.mission').length).toEqual(1);
      expect(wrapper.find('.about').length).toEqual(1);
    });

    it('should be able to start the tour', () => {
      wrapper.find('.hero__start').simulate('click');
      expect(wrapper.find('.joyride-beacon').html()).toMatchSnapshot();
      expect(wrapper.instance().joyride.props.stepIndex).toBe(0);
    });

    it('should be able to click the 1st step beacon', () => {
      wrapper.find('.joyride-beacon').simulate('click');

      Element.prototype.clientHeight = 115;
      Element.prototype.clientWidth = 450;
      expect(wrapper.find('.joyride-tooltip').html()).toMatchSnapshot();
    });

    it('should be able to close the 1st step tooltip', () => {
      wrapper.find('.joyride-tooltip__close').simulate('click');

      expect(wrapper.instance().joyride.props.stepIndex).toBe(1);
      expect(wrapper.find('.joyride-tooltip').length).toBe(0);
      expect(wrapper.find('.joyride-beacon').length).toBe(1);
    });

    it('should be able to click the 2nd step beacon', () => {
      wrapper.find('.joyride-beacon').simulate('click');

      Element.prototype.clientHeight = 115;
      Element.prototype.clientWidth = 450;
      expect(wrapper.find('.joyride-tooltip').html()).toMatchSnapshot();
    });

    it('should be able to advance to the 3rd step by clicking the button inside the hole', () => {
      wrapper.find('.mission button').simulate('click');

      expect(wrapper.instance().joyride.props.stepIndex).toBe(3);
      expect(wrapper.find('.joyride-tooltip').length).toBe(1);
    });

    it('should have skipped the missing step', () => {
      expect(mockConsole.mock.calls[0][0]).toBe('Target not mounted');
      expect(mockConsole.mock.calls[0][1]).toEqual({
        title: 'Unmounted target',
        text: 'This step tests what happens when a target is missing',
        selector: '.not-mounted'
      });
    });

    it('should be able to close the 4th step tooltip', () => {
      wrapper.find('.joyride-tooltip__close').simulate('click');

      expect(wrapper.instance().joyride.props.stepIndex).toBe(4);
      expect(wrapper.find('.joyride-tooltip').length).toBe(0);
      expect(wrapper.find('.joyride-beacon').length).toBe(1);
    });

    it('should be able to click the 5th step beacon', () => {
      wrapper.find('.joyride-beacon').simulate('click');

      Element.prototype.clientHeight = 115;
      Element.prototype.clientWidth = 450;
      expect(wrapper.find('.joyride-tooltip').html()).toMatchSnapshot();
    });

    it('should be able to close the 5th step tooltip', () => {
      wrapper.find('.joyride-tooltip__close').simulate('click');

      expect(wrapper.instance().joyride.props.stepIndex).toBe(4);
      expect(wrapper.find('.joyride-tooltip').length).toBe(0);
      expect(wrapper.find('.joyride-beacon').length).toBe(0);
    });
  });

  describe('tour with `run` set to true and `type` to "single"', () => {
    beforeAll(() => {
      wrapper = setup({
        ...props,
        run: true,
        resizeDebounce: true,
        type: 'single'
      });
    });

    afterAll(() => {
      wrapper.detach();
    });

    it('should have started with a beacon', () => {
      expect(wrapper.find('Joyride').length).toBe(1);
      expect(wrapper.find('JoyrideBeacon').length).toBe(1);
    });

    it('should be able to open the 1st tooltip', () => {
      wrapper.find('JoyrideBeacon').simulate('click');

      expect(wrapper.find('JoyrideBeacon').length).toBe(0);
      expect(wrapper.find('JoyrideTooltip').length).toBe(1);
      expect(wrapper.find('.joyride-tooltip').html()).toMatchSnapshot();
    });
  });
});
