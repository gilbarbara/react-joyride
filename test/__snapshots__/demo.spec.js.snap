// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joyride standalone tooltips should be able to trigger the 1st tooltip 1`] = `"<div class=\\"joyride-tooltip joyride-tooltip--standalone bottom joyride-tooltip--animate\\" data-target=\\".hero__tooltip\\" style=\\"position: absolute; top: -15px; left: 15px; background-color: rgba(0, 0, 0, 0.8); color: rgb(255, 255, 255); text-align: center;\\"><div class=\\"joyride-tooltip__triangle joyride-tooltip__triangle-bottom\\" style=\\"left: 5%;\\"></div><button class=\\"joyride-tooltip__close joyride-tooltip__close--header\\" data-type=\\"close\\" style=\\"color: rgb(255, 255, 255);\\"></button><div class=\\"joyride-tooltip__header\\" style=\\"color: rgb(255, 255, 255); border-color: #ff67b4;\\">The classic joyride</div><div class=\\"joyride-tooltip__main\\">Let's go on a magical tour! Just click the big orange button.</div><div class=\\"joyride-tooltip__footer\\"><button class=\\"joyride-tooltip__button joyride-tooltip__button--primary\\" data-type=\\"next\\" style=\\"color: rgb(255, 255, 255); background-color: rgb(255, 103, 180);\\">Close</button></div></div>"`;

exports[`Joyride standalone tooltips should be able to trigger the 2nd tooltip 1`] = `"<div class=\\"joyride-tooltip joyride-tooltip--standalone bottom joyride-tooltip--animate\\" data-target=\\".demo__footer img\\" style=\\"position: fixed; top: -15px; left: 15px; background-color: rgb(255, 255, 255); color: rgb(51, 51, 51); text-align: center;\\"><div class=\\"joyride-tooltip__triangle joyride-tooltip__triangle-bottom\\" style=\\"left: 5%;\\"></div><button class=\\"joyride-tooltip__close joyride-tooltip__close--header\\" data-type=\\"close\\" style=\\"color: rgb(51, 51, 51); opacity: 0;\\"></button><div class=\\"joyride-tooltip__header\\" style=\\"color: rgb(51, 51, 51);\\">A fixed tooltip</div><div class=\\"joyride-tooltip__main\\">For fixed elements, you know?</div><div class=\\"joyride-tooltip__footer\\"><button class=\\"joyride-tooltip__button joyride-tooltip__button--primary\\" data-type=\\"next\\" style=\\"color: rgb(51, 51, 51);\\">Close</button></div></div>"`;

exports[`Joyride tour with \`run\` set to false should be able to click the 1st step beacon 1`] = `"<div class=\\"joyride-tooltip bottom joyride-tooltip--animate\\" data-target=\\".projects .list\\" style=\\"position: absolute; top: -15px; left: 15px;\\"><div class=\\"joyride-tooltip__triangle joyride-tooltip__triangle-bottom\\" style=\\"left: 5%;\\"></div><button class=\\"joyride-tooltip__close joyride-tooltip__close--header\\" data-type=\\"close\\"></button><div class=\\"joyride-tooltip__header\\">Title only steps — As they say: Make the font bigger!</div><div class=\\"joyride-tooltip__main\\"></div><div class=\\"joyride-tooltip__footer\\"><button class=\\"joyride-tooltip__button joyride-tooltip__button--primary\\" data-type=\\"next\\">Next</button></div></div>"`;

exports[`Joyride tour with \`run\` set to false should be able to click the 2nd step beacon 1`] = `"<div class=\\"joyride-tooltip bottom joyride-tooltip--animate\\" data-target=\\".mission button\\" style=\\"position: absolute; top: -15px; left: 15px;\\"><div class=\\"joyride-tooltip__triangle joyride-tooltip__triangle-bottom\\" style=\\"left: 5%;\\"></div><button class=\\"joyride-tooltip__close joyride-tooltip__close--header\\" data-type=\\"close\\"></button><div class=\\"joyride-tooltip__header\\">Our Mission</div><div class=\\"joyride-tooltip__main\\">Can be advanced by clicking an element through the overlay hole.</div><div class=\\"joyride-tooltip__footer\\"><button class=\\"joyride-tooltip__button joyride-tooltip__button--secondary\\" data-type=\\"back\\">Back</button><button class=\\"joyride-tooltip__button joyride-tooltip__button--primary\\" data-type=\\"next\\" style=\\"display: none;\\">Next</button></div></div>"`;

exports[`Joyride tour with \`run\` set to false should be able to click the 5th step beacon 1`] = `"<div class=\\"joyride-tooltip bottom joyride-tooltip--animate\\" data-target=\\".demo__footer a\\" style=\\"position: fixed; top: -15px; left: 15px;\\"><div class=\\"joyride-tooltip__triangle joyride-tooltip__triangle-bottom\\" style=\\"left: 5%;\\"></div><button class=\\"joyride-tooltip__close\\" data-type=\\"close\\"></button><div class=\\"joyride-tooltip__main\\">Text only steps — Because sometimes you don't really need a proper heading</div><div class=\\"joyride-tooltip__footer\\"><button class=\\"joyride-tooltip__button joyride-tooltip__button--secondary\\" data-type=\\"back\\">Back</button><button class=\\"joyride-tooltip__button joyride-tooltip__button--primary\\" data-type=\\"next\\">Last</button></div></div>"`;

exports[`Joyride tour with \`run\` set to false should be able to start the tour 1`] = `"<button class=\\"joyride-beacon\\" style=\\"left: 15px; position: absolute; top: 15px;\\"><span class=\\"joyride-beacon__inner\\"></span><span class=\\"joyride-beacon__outer\\"></span></button>"`;

exports[`Joyride tour with \`run\` set to true and \`type\` to "single" should be able to open the 1st tooltip 1`] = `"<div class=\\"joyride-tooltip bottom joyride-tooltip--animate\\" data-target=\\".projects .list\\" style=\\"position: absolute; top: -15px; left: 15px;\\"><div class=\\"joyride-tooltip__triangle joyride-tooltip__triangle-bottom\\" style=\\"left: 5%;\\"></div><button class=\\"joyride-tooltip__close joyride-tooltip__close--header\\" data-type=\\"close\\"></button><div class=\\"joyride-tooltip__header\\">Title only steps — As they say: Make the font bigger!</div><div class=\\"joyride-tooltip__main\\"></div><div class=\\"joyride-tooltip__footer\\"><button class=\\"joyride-tooltip__button joyride-tooltip__button--primary\\" data-type=\\"close\\">Close</button></div></div>"`;
